Add a back end maintenance task for the crawler

Toflar opened the pull request:
This is a first draft and I really need some help from here on I guess :)

This is how it looks at the moment:

<img width="1198" alt="Bildschirmfoto 2019-12-04 um 18 03 03" src="https://user-images.githubusercontent.com/481937/70164037-c140c200-16c0-11ea-9939-3072724a515a.png">

ToDo's:

- [x] Bring back the FE member authentication. We need to improve this in a future version to have stateless authentication but for now, it should at least continue to work as before.
- [x] Styling. It's all just some inline styles and not really pretty. I'm just really lacking the skills here.
- [x] What configuration options should we provide? All of them according to the crawl command? Does it even make sense to let the users decide on concurrency etc.?
- [x] Provide subscriber-specific logs (for level info+)

Toflar committed:
First draft for the back end implementation

Toflar committed:
Implemented subscriber specific log files

Toflar committed:
CS

Toflar committed:
CrawlCommand CS

Toflar committed:
Discard back end configuration

Toflar committed:
Re-implemented FE member auth as previously implemented

Toflar committed:
CS

Toflar committed:
Fix issue with timeout because of session deadlock

Toflar committed:
Make sure the back end is never followed

Toflar committed:
Merge branch 'master' into feature/crawl-backend

leofeyer commented on core-bundle/src/Resources/contao/dca/tl_search_index_queue.php:
```diff
@@ -0,0 +1,60 @@
+ * @license LGPL-3.0-or-later
+ */
+
+$GLOBALS['TL_DCA']['tl_search_index_queue'] = array
```
We should probably make this a Doctrine entity (see https://github.com/contao/contao/blob/master/core-bundle/src/Entity/RememberMe.php)

Toflar replied:
Done in 3bdd8dca12624b65697917717e64481e82760bd7. I've also properly named it because it's related to crawling and not the search index at all 😏

leofeyer commented on core-bundle/src/Resources/contao/languages/en/tl_maintenance.xlf:
```diff
@@ -116,6 +92,30 @@
       <trans-unit id="tl_maintenance_jobs.symlinks.1">
         <source>Recreates the symlinks in the &lt;code&gt;web/&lt;/code&gt; folder (document root).</source>
       </trans-unit>
+      <trans-unit id="tl_maintenance.crawl.subscribersLabel.0">
```
Are you sure that four nesting levels actually work? Is it necessary to group them?

Toflar replied:
It's just an array and if it would not work you would see that there are no labels in the back end, right? 😄

leofeyer replied:
It just works accidentally because we have to support some legacy keys with a dot in the key:

https://github.com/contao/contao/blob/10f7ea76a0e3a745d65b96a86f34d512ac99186f/core-bundle/src/Config/Loader/XliffFileLoader.php#L129-L132

However, keys should not have more than three nesting levels. We even have set up a unit test to ensure we are not using more. 😄 

https://github.com/contao/contao/blob/10f7ea76a0e3a745d65b96a86f34d512ac99186f/core-bundle/tests/Fixtures/vendor/contao/test-bundle/Resources/contao/languages/en/error.xlf#L5-L8

Toflar replied:
Strange, because we're using `tl_maintenance`, not `tl_layout` here.
Anyway, fixed in 9a5a9e315260d69473b0e56ebbea21107d080641. Also provided the missing label for the broken link checker :)

leofeyer commented on core-bundle/src/Resources/contao/templates/backend/be_main.html5:
```diff
@@ -11,6 +11,7 @@
       <meta name="generator" content="Contao Open Source CMS">
       <meta name="viewport" content="width=device-width,initial-scale=1.0,shrink-to-fit=no">
       <meta name="referrer" content="origin">
+      <meta name="robots" content="noindex, nofollow">
```
Why is it necessary to add this? The `/contao` route is exempt in the `robots.txt` file and no crawler will ever be able to see this template without being logged into the back end.

Toflar replied:
You are logged into the back end if you authenticate with a front end user. There's no possibility to authenticate as a FE user statelessly at the moment.
And `/contao` is not except by the robots.txt apparently, otherwise Escargot would not crawl.

leofeyer replied:
> You are logged into the back end if you authenticate with a front end user.

What? That would be a major security hole!

Toflar replied:
You misunderstood. I mean when you rebuild the search index from the back end, you are authenticated as a back end user and front end user at the same time. That's why.

leofeyer replied:
I see.

> And `/contao` is not except by the robots.txt apparently, otherwise Escargot would not crawl.

You have to delete the `web/robots.txt` file so the request triggers the new robots.txt route. It seems we forgot to remove the file when we merged #717. 🙈 

<img width="302" alt="" src="https://user-images.githubusercontent.com/1192057/71989422-eed5d980-3231-11ea-8937-8e6dfb5f5e6e.png">

Toflar replied:
No, it was never there: https://github.com/contao/contao/pull/717/files#diff-a793700f1c87e1a3a0d01553e0c01f39
But it would make sense, yes. I'll check.

Toflar replied:
Aaah, I see what you mean now. You're right. I had to remove the file and it's all good now.
However, I don't think deleting it is a good idea. The migration would need to copy the contents into the root pages so a possible previous configuration doesn't get lost.

Toflar replied:
And I would still like to keep the meta tags because they aren't wrong :)

leofeyer committed:
Fix the coding style

Toflar committed:
Converted DCA to Doctrine entity and also properly named it

Toflar committed:
Fixed labels

leofeyer committed:
Adjust the back end implementation

leofeyer reviewed (changes requested):
I have made the necessary adjustments in aef67239b3dcc3be6c00594f01c415b946fd5319. Here are some screenshots:

<img width="737" alt="" src="https://user-images.githubusercontent.com/1192057/72072231-c6131a00-32ed-11ea-8d5c-3e8ddbddbacc.png">

<img width="790" alt="" src="https://user-images.githubusercontent.com/1192057/72072243-cca19180-32ed-11ea-8478-791ae68e1e07.png">

There are still some things which need to be fixed though:

1. The broken link checker never finishes because of an error: `SQLSTATE[22001]: String data, right truncated: 1406 Data too long for column 'uri' at row 1`.

2. A max depth of 32 seems pretty high to me. The broken link checker worked best at a max depth of 3 – anything beyond just made the process longer without improving the result.

3. The success and warning messages are in English and there is no way to translate them.

4. I am not yet happy with the "download debug log" button. If the warning message was translatable, we could integrate the link into the message, which I would prefer.

5. I still think that we do not need meta robots tags in the back end.

leofeyer commented:
Another thing that I am not happy with is using `sys_get_temp_dir()`. We have had multiple issues with it in the past (open_basedir, write permissions etc.), therefore we are only using it in unit tests.

Can we create the files in `var/` or another folder within the installation instead?

Toflar commented:
We need a temp dir that's cleaned up and `sys_get_temp_dir()` is the only one we have. Creating in `var` would never remove them and clutter the system.

Toflar commented:
Also `sys_get_temp_dir()` was created exaclty for that use case. It has to be configured correctly, that's just a hard requirement. Even in Symfony there`s loads of `sys_get_temp_dir()` calls and in other bundles likely too.

Toflar committed:
Translated summary and warnings of all subscribers for the back end

Toflar committed:
Removed robots meta tags

Toflar commented:
>     1. The broken link checker never finishes because of an error: `SQLSTATE[22001]: String data, right truncated: 1406 Data too long for column 'uri' at row 1`.

Needs to be fixed in Escargot: https://github.com/terminal42/escargot/issues/9


>     2\. A max depth of 32 seems pretty high to me. The broken link checker worked best at a max depth of 3 – anything beyond just made the process longer without improving the result.

I disagree. In my tests it actually misses quite a lot with a depth of only 3. But we might configure 8 or so. Would that be okay?

>     3\. The success and warning messages are in English and there is no way to translate them.

Fixed in 68e636af1ec5380ffbd5fd4f887eeba327839e94.

>     4\. I am not yet happy with the "download debug log" button. If the warning message was translatable, we could integrate the link into the message, which I would prefer.

Imho that log should go to the top right or so because it's a general log file and right now it looks as if it would belong to a certain subscriber.



>     5\. I still think that we do not need meta robots tags in the back end.

Removed in 68e636af1ec5380ffbd5fd4f887eeba327839e94.

Toflar committed:
CS

leofeyer committed:
Move the debug log link next to the progress bar

leofeyer committed:
Move the CRAWL. labels into the default.xlf file and fix some minor CS issues

leofeyer commented:
Thanks a lot @Toflar.

Toflar commented:
WOOOOHOOOO 🎉 🎉 🎉 🎉

xchs commented:
I am currently testing this in Contao 4.9.x-dev: should the crawler also index the "_Website root_" page (Found on level: 0)?

Toflar commented:
Sure but depends on many factors. The debug log will tell.
